8339570: Add Tidy build support for JDK tests

nizarbenalla opened the pull request:
Can I get a review for this patch that adds the necessary changes for local support of the `tidy` library.

The dependency can be retrieved by running `make/devkit/createTidyBundle.sh` on Linux and MacOs systems.

This dependency is primarily going to be used to test the generated documentation.

This patch is meant to be integrated before #21272.

Note: we need to be a very specific revision of `tidy` and cannot use any of the available artifacts, as older versions do not recognize some HTML 5 elements. 

TIA

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339570](https://bugs.openjdk.org/browse/JDK-8339570): Add Tidy build support for JDK tests (**Enhancement** - P4)


### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) Review applies to [143a84c6](https://git.openjdk.org/jdk/pull/21341/files/143a84c6e553567106c4c0858a3cd10a4f2e5e53)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Contributors
 * Magnus Ihse Bursie `<ihse@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21341/head:pull/21341` \
`$ git checkout pull/21341`

Update a local copy of the PR: \
`$ git checkout pull/21341` \
`$ git pull https://git.openjdk.org/jdk.git pull/21341/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21341`

View PR using the GUI difftool: \
`$ git pr show -t 21341`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21341.diff">https://git.openjdk.org/jdk/pull/21341.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21341#issuecomment-2395477230)

nizarbenalla commented:
Opening this PR tomorrow.
Thanks for the help on this, I waited a bit before submitting a patch while I did more testing.

/contributor add ihse

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -453,7 +453,7 @@ var getJibProfilesProfiles = function (input, common, data) {
             target_os: "macosx",
             target_cpu: "aarch64",
-            dependencies: ["devkit", "gtest", "graphviz", "pandoc"],
+            dependencies: ["devkit", "gtest", "graphviz", "pandoc", "tidy"],
```
Is there a reason for not providing Tidy on macosx-x64? It looks like the binary built by the script below would be a multi-arch variant. If so, you should deploy it as just "tidy-html-macosx" and add some code in the dependencies section below that defines "module" as just `"tidy-html" + input.target_os` on macosx.

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -1275,6 +1275,15 @@ var getJibProfilesDependencies = function (input, common) {
+            environment_name: "TIDY",
+            ext: "tar.gz",
+            revision: "5.9.20+1",
+            environment_path: input.get("tidy", "home_path") + "/tidy/bin/tidy",
+            configure_args: "TIDY=" + input.get("tidy", "home_path") +"/bin/tidy",
```
These paths are different. I'm guessing the latter one is corret?

nizarbenalla replied:
Yes, the second one is correct. Will fix it.

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -1275,6 +1275,15 @@ var getJibProfilesDependencies = function (input, common) {
         },
+        tidy: {
+            organization: common.organization,
+            environment_name: "TIDY",
```
Not sure this is a good idea. This will set an environment variable pointing to the "home_path" of the installation and will conflict with the configure arg below. The configure arg will win, so it will still work, but it could be confusing. A better environment variable name would be something like `TIDY_HOME`, but unless this is needed for something, I would just skip defining one.

nizarbenalla commented:
I made a few changes based on the review comments, I still define module as             `module: "tidy-html-" + input.target_platform` but the dependency is now provided for linux-aarch64 too

erikj79 commented:
> I made a few changes based on the review comments. After a few iterations, I now define module as `"tidy-html-" + input.target_os`. The dependency is indeed multi arch so only one is needed for macos.

When it's not too much work, we like to provide these kinds of dependencies on all platforms. Windows is obviously harder, but linux-aarch64 should be trivial. You can add a conditional, either inline using `? :` or by introducing a variable further up, and use target_os or target_platform based on if the os is macosx or not. That's what I would recommend.

magicus commented on make/RunTests.gmk:
```diff
@@ -739,6 +739,11 @@ define SetupRunJtregTestBody
   # Only the problem list for the current test root should be used.
   $1_JTREG_PROBLEM_LIST := $$(filter $$($1_TEST_ROOT)%, $$($1_JTREG_PROBLEM_LIST))
 
+  #setup tidy
```
```suggestion
  # Setup tidy
```

magicus reviewed (approved):
I think this looks overall good. Please update with the missing platforms as Erik has suggested before integrating, though.

nizarbenalla commented:
I've added the missing platform, set a `TIDY_HOME` as the enviroment_name as was suggested in the first round of review.
I've also added a new necessary condition to `make/RunTests.gmk`.

magicus commented:
The TIDY_HOME stuff is just redundant. You are setting TIDY as a configure argument, you do not need TIDY_HOME. That just adds complexity for code that will never be run. Remove the environment from the jib definition, and the check in RunTests.gmk.

nizarbenalla commented:
I've removed some of the redundant code and I've added the missing platforms.

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -1275,6 +1275,14 @@ var getJibProfilesDependencies = function (input, common) {
+            ext: "tar.gz",
+            revision: "5.9.20+1",
+            environment_path: input.get("tidy", "home_path") + "/bin/tidy",
+            configure_args: "TIDY=" + input.get("tidy", "home_path") +"/bin/tidy",
```
```suggestion
            configure_args: "TIDY=" + input.get("tidy", "home_path") + "/bin/tidy",
```

erikj79 commented on make/RunTestsPrebuiltSpec.gmk:
```diff
@@ -176,3 +176,7 @@ ULIMIT := ulimit
 ifeq ($(OPENJDK_BUILD_OS), windows)
   PATHTOOL := cygpath
 endif
+
+ifneq ($(TIDY_HOME), )
+  TIDY := $(TIDY_HOME)/bin/tidy
+endif
```
Unless TIDY_HOME is a well established variable in the context of Tidy, which I doubt, then I think this isn't needed anymore. We aren't setting this from jib-profiles.js, and I can't think of any other place this would be set. We should export `TIDY` as an environment variable in the `run-test-prebuilt-docs` profile in the other PR though.

nizarbenalla replied:
I need to push one more patch in the other PR for these to be in sync but `TIDY_HOME` is needed.

`RunTestsPrebuiltSpec.gmk` is only read if you use `make run-test-prebuilt`, without running configure. (I'm stealing Magnus' words here)

erikj79 commented on make/RunTestsPrebuiltSpec.gmk:
```diff
@@ -176,3 +176,7 @@ ULIMIT := ulimit
   PATHTOOL := cygpath
 endif
+
+ifneq ($(TIDY_HOME), )
+  TIDY := $(TIDY_HOME)
+endif
```
This is worse than before. Please forget what I said initially about `TIDY_HOME`, it should certainly not be used as an alias for `TIDY`.

nizarbenalla replied:
I may need suggestions, I pushed the last changes to the other PR [f5aeddd](https://github.com/openjdk/jdk/pull/21272/commits/f5aeddd8a2b142934558d75b5b81823ddeefd6e0).

I need a way to set `TIDY` without using configure, so that `-Dtidy` is set in RunTests.gmkl.

Maybe I can change `testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["TIDY_HOME"] = input.get("tidy", "home_path") + "/bin/tidy";` to use `TIDY` instead of `TIDY_HOME`? There would be no conflict as configure is not run and `TIDY=` was never passed

erikj79 replied:
That is exactly what I would like you to do.

magicus replied:
The two different but yet closely related PRs are making this hard to review. I thought I saw the changes to `testOnlyProfilesPrebuiltDocs` but then it was not there. Until I realized that that tidy change is in the other PR. 

That is not really ideal. Otoh, for this bug to be integrated first, it cannot rely on structs that are only created in the second PR. So I guess that is okay, for now. In the end, neither this nor the docs-test root PR will achieve much on their own, but is only laying the groundwork.

magicus reviewed (approved):
I guess that means I think this is fine. But please also wait for Erik's review.

vidmik commented on make/devkit/createTidyBundle.sh:
```diff
@@ -0,0 +1,86 @@
+
+tar -czvf "$PACKAGED_FILE" -C $INSTALL_PREFIX.. tidy
+
+echo "Created $INSTALL_PREFIX/../$PACKAGED_FILE"
```
Nit: `$INSTALL_PREFIX` already incudes a slash. I think it would be clearer if it didn't, other variables like it tend not to.

nizarbenalla replied:
Fixed in [47a67f2](https://github.com/openjdk/jdk/pull/21341/commits/47a67f2eaba3a4fe1afa1e09cdf03d334b55c2d2), thanks.

vidmik commented on make/devkit/createTidyBundle.sh:
```diff
@@ -0,0 +1,86 @@
+esac
+
+cd "$SCRIPT_DIR"
+rm -rf $SRC_DIR
```
Double quotes for good luck

vidmik commented on make/devkit/createTidyBundle.sh:
```diff
@@ -0,0 +1,86 @@
+cd "$INSTALL_PREFIX.."
+PACKAGED_FILE="tidy-html5.tar.gz"
+
+tar -czvf "$PACKAGED_FILE" -C $INSTALL_PREFIX.. tidy
```
Double quotes around `$INSTALL_PREFIX` for good luck

vidmik commented on make/autoconf/lib-tests.m4:
```diff
@@ -308,6 +308,32 @@ AC_DEFUN_ONCE([LIB_TESTS_SETUP_JIB],
+    else
+      AC_MSG_RESULT([yes])
+      AC_MSG_CHECKING([for tidy version])
+      TIDY_VERSION=`$ECHO $tidy_output | $SED -e 's/.*version //g'`
```
Looks like `TIDY_VERSION` isn't "exported", should it be lower-case?

magicus replied:
We're not really consistant on that usage. In fact, I'd say that most variables are upper case, regardless of how "local" their use is, and the lower case variant is a minority. So I'd say it's up to Nizar; if he wants to change this he can, otherwise it can stay upper case.

nizarbenalla replied:
I lowercased it just to be safe.

erikj79 commented on make/autoconf/lib-tests.m4:
```diff
@@ -308,6 +308,32 @@ AC_DEFUN_ONCE([LIB_TESTS_SETUP_JIB],
+      AC_MSG_RESULT([yes])
+      AC_MSG_CHECKING([for tidy version])
+      tidy_version=`$ECHO $tidy_output | $SED -e 's/.*version //g'`
+      AC_MSG_RESULT([$TIDY_VERSION])
```
You missed lower casing the reference here.

```suggestion
      AC_MSG_RESULT([$tidy_version])
```

nizarbenalla replied:
You're right, I missed this

nizarbenalla commented:
Thank you all for the reviews and help getting this feature in. Here goes

/integrate
